pkg/cli/admin/nodeimage/monitor.go

```diff
@@ -1,52 +1,368 @@
+// monitor cli tool. Logs from node-joiner monitor are streamed from the pod
+// to stdout.
+func (o *MonitorOptions) Run(ctx context.Context) error {
+	defer o.cleanup(ctx)
+
+	err := o.createNamespace(ctx)
+	if err != nil {
+		return nil
+	}
+
+	err = o.runNodeJoinerPod(ctx)
+	if err != nil {
+		return err
+	}
+
+	podName := o.nodeJoinerPod.GetName()
+
+	if err := o.waitForContainerRunning(ctx); err != nil {
+		klog.Errorf("monitoring did not start: %s", err)
+		return fmt.Errorf("monitoring did not start for pod %s: %s", podName, err)
+	}
+
+	if err := o.waitForMonitoringToComplete(ctx); err != nil {
+		klog.Errorf("monitoring never finished: %v", err)
+		if exiterr, ok := err.(*utilsexec.CodeExitError); ok {
+			return exiterr
+		}
+		return fmt.Errorf("monitoring never finished for pod %s: %s", podName, err)
+	}
+
+	return nil
+}
+
+func (o *MonitorOptions) waitForMonitoringToComplete(ctx context.Context) error {
+	pod, err := o.Client.CoreV1().Pods(o.nodeJoinerNamespace.Name).Get(ctx, o.nodeJoinerPod.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	opts := &logs.LogsOptions{
+		Options: &corev1.PodLogOptions{
+			// tail the pod's log instead of printing the entire log every poll
+			Follow: true,
+		},
+		RESTClientGetter: o.RESTClientGetter,
+		Object:           pod,
+		ConsumeRequestFn: logs.DefaultConsumeRequest,
+		LogsForObject:    polymorphichelpers.LogsForObjectFn,
+		IOStreams:        genericiooptions.IOStreams{Out: o.Out},
+	}
+
+	return wait.PollUntilContextTimeout(
+		ctx,
+		time.Second*5,
+		time.Minute*90,
+		true,
+		func(ctx context.Context) (done bool, err error) {
+			if err := opts.RunLogs(); err != nil {
+				return false, err
+			}
+
+			return o.isMonitoringDone(ctx)
+		})
+}
+
+func (o *MonitorOptions) isMonitoringDone(ctx context.Context) (bool, error) {
+	pod, err := o.Client.CoreV1().Pods(o.nodeJoinerNamespace.Name).Get(ctx, o.nodeJoinerPod.Name, metav1.GetOptions{})
+	if err != nil {
+		// at this stage pod should exist, so error if not found
+		if kapierrors.IsNotFound(err) {
+			return true, err
+		}
+		return false, nil
+	}
+	var state *corev1.ContainerState
+	for _, cstate := range pod.Status.ContainerStatuses {
+		if cstate.Name == nodeJoinerMonitorContainer {
+			state = &cstate.State
+			break
+		}
+	}
+
+	// missing status for monitor container => timeout in the worst case
+	if state == nil {
+		return false, nil
+	}
+
+	if state.Terminated != nil {
+		if state.Terminated.ExitCode == 0 {
+			return true, nil
+		}
+		return true, &utilsexec.CodeExitError{
+			Err:  fmt.Errorf("%s/%s unexpectedly terminated: exit code: %v, reason: %s, message: %s", o.nodeJoinerNamespace.Name, o.nodeJoinerPod.Name, state.Terminated.ExitCode, state.Terminated.Reason, state.Terminated.Message),
+			Code: int(state.Terminated.ExitCode),
+		}
+	}
+	return false, nil
+}
+
+func (o *MonitorOptions) runNodeJoinerPod(ctx context.Context) error {
```

Comment:
This method (except the `tasks` list`) could be probably be defined in the base class